db/inlineskiplist.h

```diff
@@ -549,43 +665,103 @@ void InlineSkipList<Comparator>::Insert(const char* key) {
+        KeyIsNextToNode(key, hint->prev[level - 1], level - 1)) {
+      level--;
     }
-
-    // It is ok to mutate max_height_ without any synchronization
-    // with concurrent readers.  A concurrent reader that observes
-    // the new value of max_height_ will see either the old value of
-    // new level pointers from head_ (nullptr), or a new value set in
-    // the loop below.  In the former case the reader will
-    // immediately drop to the next level since nullptr sorts after all
-    // keys.  In the latter case the reader will use the new node.
-    max_height_.store(height, std::memory_order_relaxed);
   }
+  if (level > stop_level) {
+    FindLessThan(key, hint->prev, hint->prev[level], level, stop_level);
+  }
+}
 
-  for (int i = 0; i < height; i++) {
-    // NoBarrier_SetNext() suffices since we will add a barrier when
-    // we publish a pointer to "x" in prev[i].
-    x->NoBarrier_SetNext(i, prev_[i]->NoBarrier_Next(i));
-    prev_[i]->SetNext(i, x);
+template <class Comparator>
+void InlineSkipList<Comparator>::InsertWithHint(
+    const char* key, void** hint_ptr) {
+  int height = 0;
+  Node* x = GetNodeForInsert(key, &height);
+  MaybeInvalidatePrev(height);
+  
+  assert(hint_ptr != nullptr);
+  InsertHint* hint = reinterpret_cast<InsertHint*>(*hint_ptr);
+  if (hint == nullptr) {
+    hint = AllocateInsertHint();
+    *hint_ptr = reinterpret_cast<void*>(hint);
+  }
+  int level = 0;
+  for (; level < hint->num_levels; level++) {
+    if (KeyIsAfterNode(key, hint->prev[level])) {
+      break;
+    }
+  }
+  if (level >= hint->num_levels) {
+    FindWithHint(key, hint, std::max<int>(level, height - 1), level);
+  } else {
+    int hint_max_height = 0;
+    for (int i = level; i < hint->num_levels; i++) {
+      hint_max_height = std::max<int>(hint_max_height, hint->prev_height[i]);
+    }
+    if (height > hint_max_height) {
+      FindWithHint(key, hint, height - 1, hint_max_height);
+    }
+  }
+  int current_level = 0;
+  if (level > 0) {
+    Node* tmp[level];
+    FindLessThan(key, tmp, hint->prev[level], level, 0);
+    while (current_level < height && current_level < level) {
+      x->InsertAfter(tmp[current_level], current_level);
+      current_level++;
+    }
+  }
+  for (int i = level; i < hint->num_levels; i++) {
+    while (current_level < height && current_level < hint->prev_height[i]) {
+      x->InsertAfter(hint->prev[i], current_level);
+      current_level++;
+    }
+  }
+  while (current_level < height) {
+    x->InsertAfter(hint->prev[current_level], current_level);
+    current_level++;
+  }
+  Node* p = x;
+  uint8_t h = static_cast<uint8_t>(height);
+  for (int i = level; i < hint->num_levels; i++) {
+    if (h <= i) {
+      p = nullptr;
+      break;
+    }
+    std::swap(p, hint->prev[i]);
+    std::swap(h, hint->prev_height[i]);
+  }
+  if (p != nullptr && h > hint->num_levels) {
+    hint->prev[hint->num_levels] = p;
+    hint->prev_height[hint->num_levels] = h;
+    hint->num_levels++;
```

Comment:
This function is too hard for me to understand. Is there a way we go with a simple solution that treats hint->prev the same way as prev_?